src/event_reader.rs

```diff
@@ -51,40 +67,46 @@ pub type SegmentReadResult = Result<SegmentDataBuffer, ReaderError>;
 ///         scope: Scope::from("scope".to_string()),
 ///         stream: Stream::from("stream".to_string()),
 ///     };
-///     // Create a reader obtain a segment slice and read events from it.
-///     let mut reader = client_factory.create_event_stream_reader(stream).await;
+///     // Create a reader group to read data from the Pravega stream.
+///     let rg = client_factory.create_reader_group("rg".to_string(), stream).await;
+///     // Create a reader under the reader group. The segments of the stream are assigned among the
+///     // readers which are part of the reader group.
+///     let mut reader1 = rg.create_reader("r1".to_string()).await;
 ///     // read all events from a given segment slice.
-///     if let Some(mut segment_slice) =  reader.acquire_segment().await {
+///     if let Some(mut segment_slice) =  reader1.acquire_segment().await {
 ///         while let Some(event) = segment_slice.next() {
 ///             println!("Event read is {:?}", event);
 ///         }
 ///     }
 ///     // read one event from the a given  segment slice and return it back.
-///     if let Some(mut segment_slice) = reader.acquire_segment().await {
+///     if let Some(mut segment_slice) = reader1.acquire_segment().await {
 ///         if let Some(event) = segment_slice.next() {
 ///             println!("Event read is {:?}", event);
 ///             // release the segment slice back to the reader.
-///             reader.release_segment_at(segment_slice);
+///             reader1.release_segment(segment_slice);
 ///         }
 ///     }
 /// }
-///```
+/// ```
 ///
 #[derive(new)]
 pub struct EventReader {
-    stream: ScopedStream,
+    id: Reader,
     factory: ClientFactory,
     rx: Receiver<SegmentReadResult>,
     tx: Sender<SegmentReadResult>,
     meta: ReaderMeta,
+    rg_state: Arc<Mutex<ReaderGroupState>>,
 }
 
 /// Reader meta data.
 pub struct ReaderMeta {
     slices: HashMap<String, SliceMetadata>,
-    future_segments: HashMap<ScopedSegment, HashSet<Segment>>,
-    slice_release_receiver: HashMap<String, oneshot::Receiver<SliceMetadata>>,
+    slices_dished_out: HashMap<String, i64>,
+    slice_release_receiver: HashMap<String, oneshot::Receiver<Option<SliceMetadata>>>,
     slice_stop_reading: HashMap<String, oneshot::Sender<()>>,
+    last_segment_release: Instant,
```

Comment:
Perhaps I'm missing something but I don't quite understand how rebalance works. Looks like it checks the `last_segment_release` time to see if there are any slice release in the past 10 seconds to decide if it needs to unassign the segment. But I don't see this `last_segment_release` updated anywhere except in the init method. Does that mean it will always rebalance after 10 seconds of the `EventReader` init?

Reply:
Thanks, fixed the issue. This issue was present even on the acquire_segment path too. Added a junit for acquire test.